tests/e2e/preferential_topology_utils.go

```diff
@@ -0,0 +1,456 @@
+	counter := 0
+	stsPodCount := 0
+	var isDatastoreMatchFound bool
+	var ssPodsBeforeScaleDown *v1.PodList
+	if parallelStatefulSetCreation {
+		ssPodsBeforeScaleDown = GetListOfPodsInSts(client, statefulset)
+	} else {
+		ssPodsBeforeScaleDown = fss.GetPodList(client, statefulset)
+	}
+	stsPodCount = len(ssPodsBeforeScaleDown.Items)
+	for _, sspod := range ssPodsBeforeScaleDown.Items {
+		_, err := client.CoreV1().Pods(namespace).Get(ctx, sspod.Name, metav1.GetOptions{})
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+		for _, volumespec := range sspod.Spec.Volumes {
+			if volumespec.PersistentVolumeClaim != nil {
+				pv := getPvFromClaim(client, statefulset.Namespace, volumespec.PersistentVolumeClaim.ClaimName)
+				ginkgo.By(fmt.Sprintf("Invoking QueryCNSVolumeWithResult with VolumeID: %s",
+					pv.Spec.CSI.VolumeHandle))
+				for i := 0; i < len(datastoreNames); i++ {
+					if val, ok := datastoreListMap[datastoreNames[i]]; ok {
+						isDatastoreMatchFound = e2eVSphere.verifyPreferredDatastoreMatch(pv.Spec.CSI.VolumeHandle, val)
+						if isDatastoreMatchFound {
+							framework.Logf("volume %s is created on the chosen preferred datastore %v",
+								pv.Spec.CSI.VolumeHandle, val)
+							counter = counter + 1
+							break
+						}
+					}
+				}
+			}
+		}
+	}
+	if len(datastoreNames) == 1 && counter != stsPodCount && !multipleAllowedTopology {
+		return fmt.Errorf("volume is provisioned on the wrong datastore")
+	} else if len(datastoreNames) == 2 && counter != stsPodCount && !multipleAllowedTopology {
+		return fmt.Errorf("volume is provisioned on the wrong datastore")
+	} else if len(datastoreNames) == 3 && counter != stsPodCount && !multipleAllowedTopology {
+		return fmt.Errorf("volume is provisioned on the wrong datastore")
+	} else if len(datastoreNames) == 4 && counter != stsPodCount && !multipleAllowedTopology {
+		return fmt.Errorf("volume is provisioned on the wrong datastore")
+	} else if counter != stsPodCount && multipleAllowedTopology && counter != 0 {
+		framework.Logf("Few volume is provisioned on some other datastore due to multiple " +
+			"allowed topology set or no topology set in the Storage Class")
+	}
+	return nil
+}
+
+/*
+	verifyVolumeProvisioningForStandalonePods is used to check whether the volume is provisioned on the
+
+chosen preferred datastore or not for standalone pods
+*/
+func verifyVolumeProvisioningForStandalonePods(ctx context.Context,
+	client clientset.Interface, pod *v1.Pod,
+	namespace string, datastoreNames []string, datastoreListMap map[string]string) {
+	var flag bool = false
+	for _, volumespec := range pod.Spec.Volumes {
+		if volumespec.PersistentVolumeClaim != nil {
+			pv := getPvFromClaim(client, pod.Namespace, volumespec.PersistentVolumeClaim.ClaimName)
+			ginkgo.By(fmt.Sprintf("Invoking QueryCNSVolumeWithResult with VolumeID: %s",
+				pv.Spec.CSI.VolumeHandle))
+			for i := 0; i < len(datastoreNames); i++ {
```

Comment:
We are not calling queryCNS volume API here right?

Reply:
ginkgo message corrected